### PR TITLE
Adding NTLMSSP support

### PR DESCRIPTION
**Pull request by ropnop**
Hello! Wanted to open this to start a conversation around adding NTLMSSP bind support to go-ldap/v3

I got this working by leveraging github.com/Azure/go-ntlmssp, which can handle negotiate and response message generation

To get the messages encoded properly, I also had to make some changes to github.com/go-asn1-ber/asn1-ber, which I submitted a PR for as well (https://github.com/go-asn1-ber/asn1-ber/pull/27). Without that patch, `ber.Encode` was skipping over the negotiate and challenge-response messages.

In the end, this works on my AD domain with the following code:

```
func must(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func main() {

	l, err := ldap.DialURL("ldap://pdc01.lab.ropnop.com:389")
	if err != nil {
		log.Fatal(err)
	}
	defer l.Close()

	l.Debug = true

	err = l.NTLMBind("", "thoffman", "Password123")
	must(err)
}
```
We get a successful bind request and subsequent searches work.

Let me know what you think! I also have an idea to add support for authenticating with just an NT hash as well if you already have it, but that would require a change to the go-ntlmssp module as well.

Great library btw!

### CODE REVIEW

File: v3/bind.go [hunk 1]
```diff
@@ -6,6 +6,7 @@ import (
 	enchex "encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/Azure/go-ntlmssp"
```
Comment (johnweldon) [inline v3/bind.go:9]: nitpick - I prefer to see third party imports grouped separately from stdlib imports.  You can move this to right after the `ber` import.

File: v3/bind.go [hunk 2]
```diff
@@ -387,3 +388,118 @@ func (l *Conn) ExternalBind() error {
 
 	return GetLDAPError(packet)
 }
+
+// NTLMBind performs an NTLMSSP bind leveraging https://github.com/Azure/go-ntlmssp
+
+type NTLMBindRequest struct {
```
Comment (johnweldon) [inline v3/bind.go:394]: Exported symbols should have document comments.

File: v3/bind.go [hunk 3]
```diff
@@ -387,3 +388,118 @@ func (l *Conn) ExternalBind() error {
...
+	auth := ber.Encode(ber.ClassContext, ber.TypePrimitive, ber.TagEnumerated, negMessage, "authentication")
+	request.AppendChild(auth)
+	envelope.AppendChild(request)
+	if len(req.Controls) > 0 {
+		envelope.AppendChild(encodeControls(req.Controls))
+	}
+	return nil
+}
+
+type NTLMBindResult struct {
```
Comment (johnweldon) [inline v3/bind.go:419]: Doc comment ...

File: v3/bind.go [hunk 4]
```diff
@@ -387,3 +388,118 @@ func (l *Conn) ExternalBind() error {
...
+	Controls []Control
+}
+
+func (l *Conn) NTLMBind(domain, username, password string) error {
```
Comment (johnweldon) [inline v3/bind.go:423]: Doc comment

File: v3/bind.go [hunk 5]
```diff
@@ -387,3 +388,118 @@ func (l *Conn) ExternalBind() error {
...
+	req := &NTLMBindRequest{
+		Domain:   domain,
+		Username: username,
+		Password: password,
+	}
+	_, err := l.NTLMChallengeBind(req)
+	return err
+}
+
+func (l *Conn) NTLMChallengeBind(ntlmBindRequest *NTLMBindRequest) (*NTLMBindResult, error) {
```
Comment (johnweldon) [inline v3/bind.go:433]: Doc comment

File: v3/bind.go (hunk 3 above)
Comment (ropnop) [reply v3/bind.go:429]: lmk if that's enough details or if I should add more

**Comment by ropnop** [comment]
Thanks for the feedback! I added comments, moved the import and copied the changes to the base `bind.go`. I'm still digging into how we can add tests I'll see if I can create something at least

**Comment by johnweldon** [comment]
Looks good; when you're ready to merge you can rebase your branch on master

**Comment by ropnop** [comment]
Merged master back in if you’re ready to merge this PR. The only way I can think of when it comes to testing is either mocking out an NTLM LDAP server or spinning one up in a Docker container for integration testing. Is that something you think is worthwhile?

**Comment by johnweldon** [comment]
No, I don't want to try mocking out a server or spinning one up for the tests.

Thanks @ropnop

**Comment by johnweldon** [comment]
Released in version 3.2.0

**Pull request merged**